[feature] add stanet reproduction

sun222 opened the pull request:
压缩了提交记录，我这边看不到还有冲突问题了

sun222 committed:
Delete cdnet_predict.py

Update README.md

Create dd.l

Add files via upload

Update README.md

Update stanet.py

Create stanet.py

Create stanet.py

Add files via upload

Update README.md

Create train_infer_python.txt

Create prepare.sh

Create test_train_inference_python.sh

Create README.md

Create commen_func.sh

Rename commen_func.sh to common_func.sh

Add files via upload

Update stanet.py

Update stanet.py

Delete test_tipc/data directory

Create sss.ss

Add files via upload

Create stanet_predict.py

Update and rename stanet.py to stanet_train.py

Update and rename stanet.py to stanet_eval.py

Create stanet_infer.py

Update and rename export_model.py to stanet_export.py

Update prepare.sh

Update test_train_inference_python.sh

Delete sss.ss

Update train_infer_python.txt

Update stanet.py

Create 1.jk

Add files via upload

Delete 1.jk

Update README.md

Add files via upload

Update README.md

Update README.md

Update README.md

Update README.md

Update README.md

Add files via upload

Update README.md

Delete tutorials/predict directory

Update README.md

Update README.md

Update README.md

Update README.md

Update stanet_train.py

Update stanet_eval.py

Update stanet_infer.py

Create stanet_predict.py

Update train_infer_python.txt

Update README.md

Update README.md

Delete TIPC_result.png

Add files via upload

Add files via upload

Update README.md

Update README.md

Delete result.png

Add files via upload

Update README.md

Update README.md

Update README.md

Update README.md

Update stanet_infer.py

Update stanet_predict.py

Update stanet_train.py

Update stanet_eval.py

Delete test_tipc directory

Update create_list.py

Update spliter-cd.py

Update stanet.py

Update stanet_export.py

Delete stanet.py

Update README.md

Delete Atrain_1_3_1.png

Delete Btrain_1_3_1.png

Delete Ltrain_1_3_1.png

Delete TIFC_next.png

Delete TIFC_pre.png

Delete a.png

Delete TIFC_result.png

Delete b.png

Delete label.png

Delete result.png

Delete stanetmodel.png

Create export_model.py

Delete output/stanet/vdl_log directory

Bobholamovic commented on deploy/export/stanet_export.py:
```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
为什么要为STANet单独添加一个export文件呢？

sun222 replied:
因为使用原来的export，导出的静态模型无法推理。我分析了一下，我认为是因为loadmodel加载方法加载的是模型默认的初始化参数，所以我单独加了一个export文件指定模型的初始信息。

Bobholamovic replied:
我看了一下`stanet_export.py`，这个具体的初始化参数的修改指的是用BAM代替PAM吗？

sun222 replied:
是的

Bobholamovic replied:
那我认为这样的做法欠妥~ 因为我们不应该绕过代码中存在的问题。我建议可以参考 #64 ，解决PAM的导出问题后，使用统一的export脚本进行模型导出。

Bobholamovic commented on paddlers/custom_models/cd/stanet.py:
```diff
@@ -215,7 +214,7 @@ def forward(self, x):
         out = F.interpolate(out, scale_factor=self.ds)
         out = out + x
-        return out.reshape(out.shape[:-1] + [out.shape[-1] // 2, 2])
+        return out.reshape(tuple(out.shape[:-1]) +tuple([out.shape[-1] // 2, 2]))
```
这里强转tuple是为了？

sun222 replied:
这里强转的原因是，out.shape[:-1]在静态和动态情况不同。在动态下是list 在静态下是tuple，为了在代码上统一，都应用tuple格式

Bobholamovic commented on paddlers/custom_models/cd/stanet.py:
```diff
@@ -59,7 +59,6 @@ def __init__(self, in_channels, num_classes, att_type='BAM', ds_factor=1):
             Conv3x3(WIDTH, num_classes))
 
         self.init_weight()
```
看起来似乎有些格式问题，请问在提交commit时是否使用了pre-commit哦？

Bobholamovic commented on tools/spliter-cd.py:
```diff
@@ -0,0 +1,59 @@
+import os
```
如果是新增了划分CD数据集的工具，也需要在相应的文档中给出说明

Bobholamovic replied:
另外，`spliter-cd.py`这一文件命名风格与同目录的其它文件并不一致。出于减少重复代码的考虑，可以尝试下能否将这一功能并入已有的`spliter.py`文件中。

sun222 replied:
spliter-cd.py我看了下，我认为在功能上可以替换掉spliter.py。并且不需要使用require.txt中不存在的gdal库的功能。我改一下参数和spliter.py一样后，可以直接替换嘛

Bobholamovic replied:
可以的

Bobholamovic commented on tutorials/eval/stanet_eval.py:
```diff
@@ -0,0 +1,58 @@
+import sys
```
我建议如果在`tutorials`目录下添加了`eval`、`test`、`predict`这些子目录的话，最好也按照`train`子目录下的格式，以任务名称新建二级子目录。并且每个tutorial以模型名命名。

Bobholamovic commented on tutorials/predict/stanet_predict.py:
```diff
@@ -0,0 +1,50 @@
+import sys
```
关于设置predict教程的目的，我不是很明白。

sun222 replied:
好的 要是没用的话，这部分我就去掉了

Bobholamovic commented on tutorials/train/stanet_train.py:
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python
```
所有tutorials建议统一保持原有风格，不需要使用命令行选项配置输入。具体可参考原有的tutorials。

sun222 committed:
按要求修改代码

1、修改loadmodel 和base.py 将模型初始参数可以导出，并可以导入，可以使用默认的model_export.py
2、使用spliter-cd 替换spliter.py
3、使用pres-ubmit 格式
4、移除stanet_predict教程
5、转换教程位置，并重命名，并保持原有风格，去掉命令行选项配置风格

sun222 commented:
按照所有要求修改了代码。
在base.py中 我不理解为什么将self.init_params.params参数删掉。为了导出初始参数，我又加上了。

Bobholamovic commented on tutorials/eval/change_detection/stanet.py:
```diff
@@ -0,0 +1,36 @@
+# 可使用VisualDL查看训练指标，参考https://github.com/PaddlePaddle/paddlers/blob/develop/docs/visualdl.md
+num_classes = 2
+model = pdrs.tasks.STANet(
+    in_channels=3, num_classes=num_classes, att_type='PAM', ds_factor=1)
```
目前是否STANet的PAM和BAM都能正常导出了呢？

sun222 replied:
目前是的

Bobholamovic commented on tutorials/infer/change_detection/stanet.py:
```diff
@@ -0,0 +1,51 @@
+        warmup_iters=warmup_iters,
+        repeats=repeats,
+        transforms=eval_transforms)
+    #label_map = result[0]['label_map']
```
没有用的注释应该去除～

Bobholamovic commented on tutorials/infer/change_detection/stanet.py:
```diff
@@ -0,0 +1,51 @@
+    return (arr * 255).astype('uint8')
+
+
+infer_dir = "./inference_model"
```
命名风格方面建议和原有tutorial保持一致。此外这些路径是有效的吗？还是需要用户预先生成呢？如果需要用户经过预处理得到，应当给出预处理教程。

Bobholamovic commented on tutorials/eval/change_detection/stanet.py:
```diff
@@ -0,0 +1,36 @@
+DATA_DIR = "./dataset/"
+EVAL_FILE_LIST_PATH = os.path.join(DATA_DIR, 'val.txt')
+
+STATE_DICT_PATH = "./output/home/aistudio/output/stanet/best_model/model.pdparams"
```
因为用户如果直接跑这个脚本，可能会发现找不到checkpoint。所以建议在readme中说明一下

Bobholamovic commented on tutorials/train/change_detection/stanet.py:
```diff
@@ -0,0 +1,91 @@
+import paddle.nn as nn
+import paddle
+
+DATA_DIR = "./datasettest/"
```
所有的教程都应该从数据下载开始，能够让用户不用做额外的处理就能一键执行。当前的教程似乎都缺少数据处理步骤，而且也没有给出数据处理的相关说明。

Bobholamovic reviewed (changes requested):
需要进一步修改，请参考review意见。另外tools的部分麻烦 @geoyee review一下。

Bobholamovic commented:
> 按照所有要求修改了代码。 在base.py中 我不理解为什么将self.init_params.params参数删掉。为了导出初始参数，我又加上了。

关于这一点我也很疑惑，原先的设计可能导致导出的模型只能使用默认参数配置。稳妥起见，咱们还是请原作者解释一下设计初衷吧～ @juncaipeng

Bobholamovic commented on tools/create_list.py:
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
这个文件如果只是为变化检测任务创建file_list的话，我认为不适合直接以`create_list.py`命名。因为PaddleRS还有检测、分割等其它任务，而tools是一个全局的命名空间。

sun222 replied:
1、对于create_list.py，改名为cd_filelist.py 可以嘛，或者可以推荐一个名字嘛
2、对于spliter.py，重新实现此函数后的优点有，
（1）不改变数据集原有的文件夹结构，比如test—》a——》100.png,原有的方式只能保存在test-》100.png
（2）不依赖于require.txt中不存在的gdal库。我了解了下gdal，不能简单的pip安装，原有方案需要用户自行安装gdal。
头部注释、timer注解我可以添加进来。
3、对于3个教程部分。
（1）我看到其他教程只有train部分的，我需要上传其他部分的教程嘛
（2）关于教程从数据下载开始。原作者发布了数据集的百度网盘和谷歌网盘两个版本，不容易直接通过代码下载，我也没别的地方可以长期存放数据集。
（3）在aistudio上我写了一个公开的项目，我有写从数据集解压，split处理，creat_list处理，训练，评估，导出，推理的全流程。如果github也需要，那么这样的readme文件放在那里合适呢
（4）如果数据集要用户下载解压的话，命令行式的教程可能会更有用，用户可以指定解压的数据集位置，而不是写死的位置

Bobholamovic replied:
1. 我建议可以叫`create_cd_filelists.py`；
2. 因为gdal具有延迟读写的优势，性能更好，所以我建议可以提供两种方案，即，在代码中检查用户是否安装了gdal，如果安装了则优先使用之，否则使用当前的eager方案。
3. 你说的这些确实都是值得考虑的问题。 稳妥起见，我建议本次PR先解决其它部分，tutorial的部分我们可以另开一个PR，等大家一起讨论好整体方案后再加上～

Bobholamovic replied:
> 1. 我建议可以叫`create_cd_filelists.py`；
> 2. 因为gdal具有延迟读写的优势，性能更好，所以我建议可以提供两种方案，即，在代码中检查用户是否安装了gdal，如果安装了则优先使用之，否则使用当前的eager方案。
> 3. 你说的这些确实都是值得考虑的问题。 稳妥起见，我建议本次PR先解决其它部分，tutorial的部分我们可以另开一个PR，等大家一起讨论好整体方案后再加上～

对于第2条，如果你认可我的建议并打算这么改的话，我觉得最好要打印一下输出，提示用户目前用的是哪一种backend。

kongdebug replied:
> 1、对于create_list.py，改名为cd_filelist.py 可以嘛，或者可以推荐一个名字嘛 2、对于spliter.py，重新实现此函数后的优点有， （1）不改变数据集原有的文件夹结构，比如test—》a——》100.png,原有的方式只能保存在test-》100.png （2）不依赖于require.txt中不存在的gdal库。我了解了下gdal，不能简单的pip安装，原有方案需要用户自行安装gdal。 头部注释、timer注解我可以添加进来。 3、对于3个教程部分。 （1）我看到其他教程只有train部分的，我需要上传其他部分的教程嘛 （2）关于教程从数据下载开始。原作者发布了数据集的百度网盘和谷歌网盘两个版本，不容易直接通过代码下载，我也没别的地方可以长期存放数据集。 （3）在aistudio上我写了一个公开的项目，我有写从数据集解压，split处理，creat_list处理，训练，评估，导出，推理的全流程。如果github也需要，那么这样的readme文件放在那里合适呢 （4）如果数据集要用户下载解压的话，命令行式的教程可能会更有用，用户可以指定解压的数据集位置，而不是写死的位置

是不是论文复现要求提的PR，之前问过RD了，验收是在自己的repo上验收，后续再迁移到pdrs里，以不打乱pdrs原有的风格为主。而且好像兄弟你是fork了我的repo，看着风格很像，我是为了论文复现的验收才单独写了eval predict infer这些代码的，具体合并什么代码或许可以再沟通，不用全部合并的吧

sun222 replied:
> > 1、对于create_list.py，改名为cd_filelist.py 可以嘛，或者可以推荐一个名字嘛 2、对于spliter.py，重新实现此函数后的优点有， （1）不改变数据集原有的文件夹结构，比如test—》a——》100.png,原有的方式只能保存在test-》100.png （2）不依赖于require.txt中不存在的gdal库。我了解了下gdal，不能简单的pip安装，原有方案需要用户自行安装gdal。 头部注释、timer注解我可以添加进来。 3、对于3个教程部分。 （1）我看到其他教程只有train部分的，我需要上传其他部分的教程嘛 （2）关于教程从数据下载开始。原作者发布了数据集的百度网盘和谷歌网盘两个版本，不容易直接通过代码下载，我也没别的地方可以长期存放数据集。 （3）在aistudio上我写了一个公开的项目，我有写从数据集解压，split处理，creat_list处理，训练，评估，导出，推理的全流程。如果github也需要，那么这样的readme文件放在那里合适呢 （4）如果数据集要用户下载解压的话，命令行式的教程可能会更有用，用户可以指定解压的数据集位置，而不是写死的位置
> 
> 是不是论文复现要求提的PR，之前问过RD了，验收是在自己的repo上验收，后续再迁移到pdrs里，以不打乱pdrs原有的风格为主。而且好像兄弟你是fork了我的repo，看着风格很像，我是为了论文复现的验收才单独写了eval predict infer这些代码的，具体合并什么代码或许可以再沟通，不用全部合并的吧

是论文复现要求提的PR。你项目做的比我快，我很多功能参考了你的代码，我在项目的readme上写了这一点。我不知道要求是先验收后合并。我不明白你的意思，是先把合并放一放，还是先把tutorial放一放。

kongdebug replied:
我认为是tutorial里的eval infer predict可以放一放，这三个文件是我为了验收方便写的，有可能和pdrs之前就有的接口重复了。还有就是数据方面，为了比赛的验收，加上对变化检测的数据切片、生成txt文件的代码，方便验收的RD全流程走通，但是这部分的代码是否加到pdrs来，还请 @Bobholamovic 漫晖哥来看看

Bobholamovic replied:
> 我认为是tutorial里的eval infer predict可以放一放，这三个文件是我为了验收方便写的，有可能和pdrs之前就有的接口重复了。还有就是数据方面，为了比赛的验收，加上对变化检测的数据切片、生成txt文件的代码，方便验收的RD全流程走通，但是这部分的代码是否加到pdrs来，还请 @Bobholamovic 漫晖哥来看看

嗯嗯，我同意tutorial部分先放一放，等和大佬们讨论好之后再提另外的PR。比赛相关的一些数据处理之类的，我觉得如果能做成通用的功能的话，也可以加到RS里，但如果是那种很针对某个数据集的专用的处理我觉得就不适合加进来了。

Bobholamovic commented on tools/spliter.py:
```diff
@@ -1,50 +1,64 @@
-# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
为什么去掉了头部注释？

Bobholamovic commented on tools/spliter.py:
```diff
@@ -1,50 +1,64 @@
+    return shotname, extension
 
 
-@Timer
```
为什么去掉`Timer`？

Bobholamovic commented on tools/spliter.py:
```diff
@@ -1,50 +1,64 @@
-                image_name + "_" + str(r) + "_" + str(c) + ".png"))
-            title.save(save_path, "PNG")
-            print("-- {:d}/{:d} --".format(int(r * cols + c + 1), total_number))
+    img_ext = [".jpg", ".png"]
```
重新实现此函数的目的是？重构的版本使用了eager方式将图像载入内存，实用性上似乎不如原先的版本。

Bobholamovic commented on tutorials/train/change_detection/stanet.py:
```diff
@@ -1,88 +1,91 @@
     early_stop=False,
     # 是否启用VisualDL日志功能
     use_vdl=True,
+    # pretrain_weights=None,
```
去除没有实际用到且不具备说明作用的注释，特别是在教程中，因为这可能让用户疑惑。

Bobholamovic commented on tutorials/train/change_detection/stanet.py:
```diff
@@ -1,88 +1,91 @@
+DATA_DIR = "./datasettest/"
+TRAIN_FILE_LIST_PATH = os.path.join(DATA_DIR, 'train.txt')
+EVAL_FILE_LIST_PATH = os.path.join(DATA_DIR, 'val.txt')
+TESTLE_LIST_PATH = os.path.join(DATA_DIR, 'test.txt')
```
`TESTLE`是否为typo？

Bobholamovic commented:
这次的PR代码量比较大，所以提的建议也稍微多了一些，辛苦同学啦！感谢贡献PR👍🏻

sun222 committed:
去掉不提交的内容

本次pr不提交这部分内容

Bobholamovic reviewed (approved):
LGTM.

geoyee commented:
> 需要进一步修改，请参考review意见。另外tools的部分麻烦 @geoyee review一下。

才看到，不好意思晖哥~关于这个spliter的改进，我有几个看法哈：
1. 现在改了之后是只能读取`.jpg`和`.png`吗，但实际上需要划分的数据应该很大部分是直接下载的`.tif`一类的。参考[GDAL支持的栅格格式](https://www.osgeo.cn/gdal/drivers/raster/index.html)，可以看到GDAL除了`.tif`等遥感常用格式外，也支持`.jpg`和`.png`的。同时统一读取再分块和直接分块读取的性能还是不一样的。且GDAL的划分支持多光谱数据，PIL应该不支持吧？
2. GDAL安装确实无法直接通过pip，但可以方便的通过whl包进行安装。[window下载](https://www.lfd.uci.edu/~gohlke/pythonlibs/#gdal)，[linux下载](https://aistudio.baidu.com/aistudio/datasetdetail/136010)。
3. 关于晖哥 @Bobholamovic 提到的判断之后选择backend为GDAL或者PIL以及大佬 @sun222 提到的“不改变数据集原有的文件夹结构”，这两点我觉得可以讨论修改加上。

Bobholamovic committed:
Merge branch 'develop' into forpr3
